### PR TITLE
Retry deadlocks on invocation update

### PR DESCRIPTION
**Pull request by tempoz**
For now, arbitrarily retries 5 times with exponential backoff.

---

**Version bump**: Patch

### CODE REVIEW

**Comment by tempoz** [comment]
Current dependencies on/for this PR:
* master
  * **PR #2090** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/2090" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/2090?utm_source=stack-comment).

**Comment by tempoz** [comment]
Addresses buildbuddy-io/buildbuddy-internal/issues#1309

File: server/backends/invocationdb/invocationdb.go [hunk 1]
```diff
@@ -108,11 +109,19 @@ func (d *InvocationDB) CreateInvocation(ctx context.Context, ti *tables.Invocati
 // id and attempt number. It returns whether a row was updated.
 func (d *InvocationDB) UpdateInvocation(ctx context.Context, ti *tables.Invocation) (bool, error) {
 	updated := false
-	err := d.h.TransactionWithOptions(ctx, db.Opts().WithQueryName("update_invocation"), func(tx *db.DB) error {
-		result := tx.Where("`invocation_id` = ? AND `attempt` = ?", ti.InvocationID, ti.Attempt).Updates(ti)
-		updated = result.RowsAffected > 0
-		return result.Error
-	})
+	var err error
+	for retries, max_retries := 0, 5; retries < 5; retries++ {
```
Comment (tylerwilliams) [inline server/backends/invocationdb/invocationdb.go:113]: maybe using util/retry here would be better?

Comment (tempoz) [reply server/backends/invocationdb/invocationdb.go:113]: oh, that's really nice.

Comment (tempoz) [reply server/backends/invocationdb/invocationdb.go]: Done!

File: server/backends/invocationdb/invocationdb.go [hunk 2]
```diff
@@ -108,11 +110,20 @@ func (d *InvocationDB) CreateInvocation(ctx context.Context, ti *tables.Invocati
...
 func (d *InvocationDB) UpdateInvocation(ctx context.Context, ti *tables.Invocation) (bool, error) {
 	updated := false
-	err := d.h.TransactionWithOptions(ctx, db.Opts().WithQueryName("update_invocation"), func(tx *db.DB) error {
-		result := tx.Where("`invocation_id` = ? AND `attempt` = ?", ti.InvocationID, ti.Attempt).Updates(ti)
-		updated = result.RowsAffected > 0
-		return result.Error
-	})
+	var err error
+	retryOptions := &retry.Options{InitialBackoff: time.Second, Multiplier: 2.0, MaxRetries: 5}
+	for i, r := 1, retry.New(ctx, retryOptions); r.Next(); i++ {
```
[the remark this answers is not in the document]
Comment (tylerwilliams) [reply server/backends/invocationdb/invocationdb.go:115]: we could add a AttemptString() string method to the retrier that returns "%d of %d" that might satisfy the logging you're doing on L122

[the remark this answers is not in the document]
Comment (tempoz) [reply server/backends/invocationdb/invocationdb.go:115]: The default doesn't have a maximum number of retries, which I'm not normally in favor of. Is it supposed to max out after it hits the max backoff? right now it just retries forever at the max backoff.

[the remark this answers is not in the document]
Comment (tylerwilliams) [reply server/backends/invocationdb/invocationdb.go:115]: seems like a bug based on the comment there:

```
MaxRetries:     0, // unlimited, time based max by default.
```

[the remark this answers is not in the document]
Comment (tempoz) [reply server/backends/invocationdb/invocationdb.go:115]: I'll fix the bug in a later PR.

**Pull request merged**